src/rules/noStaticOnlyClassesRule.ts

```diff
@@ -0,0 +1,83 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+import { getChildOfKind,
+        hasModifier,
+        isClassDeclaration,
+        isConstructorDeclaration,
+        isParameterProperty } from "tsutils";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-static-only-classes",
+        description: Lint.Utils.dedent`
+            Disallows classes containing only static members. Classes
+            with non-empty constructors are ignored.`,
+        rationale: Lint.Utils.dedent`
+            Users who come from a Java-style OO language may wrap
+            their utility functions in an extra class, instead of
+            putting them at the top level.`,
+        optionsDescription: "Not configurable.",
+        options: null,
+        optionExamples: [true],
+        type: "functionality",
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING = "Classes containing only static members are disallowed.";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new NoStaticOnlyClassesWalker(sourceFile, this.ruleName, this.ruleArguments));
+    }
+}
+
+class NoStaticOnlyClassesWalker extends Lint.AbstractWalker<string[]> {
+    public walk(sourceFile: ts.SourceFile) {
+        const checkIfStaticOnlyClass = (node: ts.Node): void => {
+            if (isClassDeclaration(node) && !isEmptyClass(node) && !hasExtendsClause(node)) {
+                for (const member of node.members) {
+                    if (isConstructorWithShorthandProps(member) ||
+                       (!isConstructorDeclaration(member) && !hasModifier(member.modifiers, ts.SyntaxKind.StaticKeyword))) {
+                        return;
+                    }
+                }
+                this.addFailureAtNode(getChildOfKind(node, ts.SyntaxKind.ClassKeyword, this.sourceFile)!, Rule.FAILURE_STRING);
+            }
+            return ts.forEachChild(node, checkIfStaticOnlyClass);
+        };
+        ts.forEachChild(sourceFile, checkIfStaticOnlyClass);
+    }
+}
+
+function hasExtendsClause(declaration: ts.ClassDeclaration): boolean {
+    return (declaration.heritageClauses !== undefined) && (declaration.heritageClauses[0].token === ts.SyntaxKind.ExtendsKeyword);
+}
+
+/**
+ * An "empty class" for our purposes.
+ */
+function isEmptyClass(declaration: ts.ClassDeclaration): boolean {
+    return declaration.members.length === 0 || declaration.members.every(isConstructorDeclaration);
```

Comment:
That check makes sense to correctly handle constructor overloads.
On a side note, you could remove the length check here because `Array.prototype.every` returns `true` for empty arrays.